File: android/a-kart/app/src/main/java/com/frogdesign/akart/Discovery.java
package com.frogdesign.akart;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;

import com.parrot.arsdk.ardiscovery.ARDiscoveryDeviceService;
import com.parrot.arsdk.ardiscovery.ARDiscoveryService;
import com.parrot.arsdk.ardiscovery.receivers.ARDiscoveryServicesDevicesListUpdatedReceiver;
import com.parrot.arsdk.ardiscovery.receivers.ARDiscoveryServicesDevicesListUpdatedReceiverDelegate;

import java.util.List;

import rx.Observable;
import rx.Subscriber;
import rx.functions.Action1;

/**
 * Created by emanuele.di.saverio on 17/12/15.
 */
public class Discovery {
    private static String TAG = Discovery.class.getSimpleName();

    private Context ctx;
    private Subscriber<? super Discovery> currentSubscriber;
    private ServiceConnection mArdiscoveryServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.i(TAG, "nexted" + currentSubscriber);
            mArdiscoveryService = ((ARDiscoveryService.LocalBinder) service).getService();
            if (currentSubscriber != null) {
                currentSubscriber.onNext(Discovery.this);
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mArdiscoveryService = null;
        }
    };
    private ARDiscoveryService mArdiscoveryService;
    private ARDiscoveryServicesDevicesListUpdatedReceiver
            mArdiscoveryServicesDevicesListUpdatedReceiver;

    public Discovery(Context ctx) {
        this.ctx = ctx.getApplicationContext();
    }

    public Observable<Discovery> binder() {
        return Observable.create(new Observable.OnSubscribe<Discovery>() {
            @Override
            public void call(Subscriber<? super Discovery> subscriber) {
                Log.i(TAG, "Subscribed on binder " + mArdiscoveryService);
                // if the discoverer service doesn't exists, bind to it
                if (mArdiscoveryService == null) {
                    currentSubscriber = subscriber;
                    Intent i = new Intent(ctx, ARDiscoveryService.class);
                    ctx.bindService(i, mArdiscoveryServiceConnection, Context.BIND_AUTO_CREATE);
                } else {
                    subscriber.onNext(Discovery.this);
                    subscriber.onCompleted();
                }
            }
        });
    }

    private Observable.OnSubscribe<List<ARDiscoveryDeviceService>> discoverer =
            new Observable.OnSubscribe<List<ARDiscoveryDeviceService>>() {
                @Override
                public void call(final Subscriber<? super List<ARDiscoveryDeviceService>> subscriber) {
                    unbind();
                    binder().subscribe(new Action1<Discovery>() {
                        @Override
                        public void call(Discovery discovery) {
                            Log.i(TAG, "nexted call");
                            ARDiscoveryServicesDevicesListUpdatedReceiverDelegate delegate =
                                    new ARDiscoveryServicesDevicesListUpdatedReceiverDelegate() {
                                        @Override
                                        public void onServicesDevicesListUpdated() {
                                            Log.d(TAG, "onServicesDevicesListUpdated ...");
                                            if (mArdiscoveryService != null) {
                                                List<ARDiscoveryDeviceService> deviceList =
                                                        mArdiscoveryService.getDeviceServicesArray();
                                                subscriber.onNext(deviceList);
                                            }
                                        }
                                    };
                            //first update is immediate
                            delegate.onServicesDevicesListUpdated();
                            mArdiscoveryServicesDevicesListUpdatedReceiver =
                                    new ARDiscoveryServicesDevicesListUpdatedReceiver(delegate);
                            LocalBroadcastManager localBroadcastMgr = LocalBroadcastManager.getInstance(ctx);
                            localBroadcastMgr.registerReceiver(mArdiscoveryServicesDevicesListUpdatedReceiver,
                                    new IntentFilter(
                                            ARDiscoveryService.kARDiscoveryServiceNotificationServicesDevicesListUpdated)

                            );
                            mArdiscoveryService.start();
                        }
                    });
                }
            };

    public Observable<List<ARDiscoveryDeviceService>> discoverer() {
        return Observable.create(discoverer);
    }

    public void unbind() {
        unregisterReceivers();
        Log.d(TAG, "closeServices ...");
        if (mArdiscoveryService != null) {
            ctx.unbindService(mArdiscoveryServiceConnection);
            mArdiscoveryService.stop();
        }
        mArdiscoveryService = null;
    }

    private void unregisterReceivers() {
        LocalBroadcastManager localBroadcastMgr = LocalBroadcastManager.getInstance(ctx);
        localBroadcastMgr.unregisterReceiver(mArdiscoveryServicesDevicesListUpdatedReceiver);
    }
}

File: android/a-kart/app/src/main/kotlin/com/frogdesign/akart/Discovery.kt
package com.frogdesign.akart

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.ServiceConnection
import android.os.IBinder
import android.support.v4.content.LocalBroadcastManager
import android.util.Log

import com.parrot.arsdk.ardiscovery.ARDiscoveryDeviceService
import com.parrot.arsdk.ardiscovery.ARDiscoveryService
import com.parrot.arsdk.ardiscovery.receivers.ARDiscoveryServicesDevicesListUpdatedReceiver
import com.parrot.arsdk.ardiscovery.receivers.ARDiscoveryServicesDevicesListUpdatedReceiverDelegate

import rx.Observable
import rx.Subscriber
import rx.functions.Action1
import timber.log.Timber

class Discovery(ctx: Context) {

    private val ctx: Context
    private var currentSubscriber: Subscriber<in Discovery>? = null
    private val mArdiscoveryServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Timber.tag(TAG).i("nexted" + currentSubscriber!!)
            mArdiscoveryService = (service as ARDiscoveryService.LocalBinder).service
            if (currentSubscriber != null) {
                currentSubscriber!!.onNext(this@Discovery)
            }
        }

        override fun onServiceDisconnected(name: ComponentName) {
            mArdiscoveryService = null
        }
    }
    private var mArdiscoveryService: ARDiscoveryService? = null
    private var mArdiscoveryServicesDevicesListUpdatedReceiver: ARDiscoveryServicesDevicesListUpdatedReceiver? = null

    init {
        this.ctx = ctx.applicationContext
    }

    fun binder(): Observable<Discovery> {
        return Observable.create { subscriber ->
            Timber.i("Subscribed on binder " + mArdiscoveryService)
            // if the discoverer service doesn't exists, bind to it
            if (mArdiscoveryService == null) {
                currentSubscriber = subscriber
                val i = Intent(ctx, ARDiscoveryService::class.java)
                ctx.bindService(i, mArdiscoveryServiceConnection, Context.BIND_AUTO_CREATE)
            } else {
                subscriber.onNext(this@Discovery)
                subscriber.onCompleted()
            }
        }
    }

    private val discoverer = Observable.OnSubscribe<List<com.parrot.arsdk.ardiscovery.ARDiscoveryDeviceService>> {
        subscriber ->
        unbind()
        binder().subscribe(rx.functions.Action1<com.frogdesign.akart.Discovery> {
            Timber.tag(TAG).i("nexted call")
            val delegate = com.parrot.arsdk.ardiscovery.receivers.ARDiscoveryServicesDevicesListUpdatedReceiverDelegate {
                Timber.tag(TAG).i( "onServicesDevicesListUpdated ...")
                if (mArdiscoveryService != null) {
                    val deviceList = mArdiscoveryService!!.deviceServicesArray
                    subscriber.onNext(deviceList)
                }
            }
            //first update is immediate
            delegate.onServicesDevicesListUpdated()
            mArdiscoveryServicesDevicesListUpdatedReceiver = ARDiscoveryServicesDevicesListUpdatedReceiver(delegate)
            val localBroadcastMgr = LocalBroadcastManager.getInstance(ctx)
            localBroadcastMgr.registerReceiver(mArdiscoveryServicesDevicesListUpdatedReceiver,
                    IntentFilter(
                            ARDiscoveryService.kARDiscoveryServiceNotificationServicesDevicesListUpdated))
            mArdiscoveryService!!.start()
        })
    }

    fun discoverer(): Observable<List<ARDiscoveryDeviceService>> {
        return Observable.create(discoverer)
    }

    fun unbind() {
        unregisterReceivers()
        Timber.tag(TAG).i( "closeServices ...")
        if (mArdiscoveryService != null) {
            ctx.unbindService(mArdiscoveryServiceConnection)
            mArdiscoveryService!!.stop()
        }
        mArdiscoveryService = null
    }

    private fun unregisterReceivers() {
        val localBroadcastMgr = LocalBroadcastManager.getInstance(ctx)
        localBroadcastMgr.unregisterReceiver(mArdiscoveryServicesDevicesListUpdatedReceiver)
    }

    companion object {
        private val TAG = Discovery::class.java.simpleName
    }
}